docs/developer/architecture/coding-principles.md

```diff
@@ -27,26 +27,64 @@
+10. use package-private classes if they're not needed outside the package, e.g. implementation classes
+11. avoid using `enum`s for anything other than named integer enumerations.
+12. avoid using static classes as much as possible. Exceptions to this are helper functions and test utils, etc.
+13. avoid any kind of variable prefixes, no underscores, no `m_` for members, etc.
+14. avoid unnecessary `this.` (exception: in builders)
+15. use static imports, as long as it is still clear what's being done. For example,
+    use `of(item1, item2).map(it -> it.someOperation)...` instead of `Stream.of(item1, item2)`, or `format("...",arg1)`
+    instead of `String.format("...", arg1)`. Helps with clarity and readability.
+16. avoid `Optional` as method return type or method argument. Use `null` in signatures.
+17. avoid cryptic variable names, especially in long methods. Instead, try to write them out, at least to a reasonable
+    extent.
+
+### V. Testing
 
 1. All handlers and services should have dedicated unit tests with mocks used for dependencies.
-2. When appropriate, prefer composing services via the constructor so that dependencies can be mocked as opposed to
-   instantiating dependencies directly, see `ProvisionManagerImpl`.
-
-### V. Data Objects
-
-1. Do not create constructors that take many parameters; instead use the builder pattern. An example
-   is `TransferProcess`.
-2. Note that the motivation behind use of builders is not for immutability (although that may be good in certain
+2. Prefer unit tests over all other test types: unit > integration/component > e2e
+3. When appropriate, prefer composing services via the constructor so that dependencies can be mocked as opposed to
+   instantiating dependencies directly.
+4. Use classes with static test functions to provide common helper methods, e.g. to instantiate an object.
+5. Use `[METHOD]_when[CONDITION]_should[EXPECTATION]` as naming template for test methods,
+   e.g. `verifyInput_whenNull_throwsNpe()` as opposed to `testInputNull()`
+
+### VI. Data Objects
+
+1. Use the `Builder` pattern when:
+    - there are any number of optional constructor args
+    - there are more than 3 constructor args
+    - the object should be serializable
+    - inheriting from an object that fulfills any of the above. In this case use derived builders as well.
+2. Builders should be named just `Builder` and be static nested classes.
+3. Create a `public static Builder newInstance(){...}` method to instantiate the builder
+4. Builders have non-public constructors
+5. use single-field builders: a `Builder` instantiates the object it builds in its constructor, and sets the properties
+   in its builder methods. The `build()` method then only performs verification (optional) and returns the instance.
+6. use `private` constructors for the objects that the builder builds.
+7. use Jackson annotations on builders only, as opposed to: the object.
```

Comment:
Jackson annotations are still required on the object, e.g. type info

Reply:
I meant annotations to configure ser/des. Theoretically users could put `@JsonProperty` annotations etc. in several places, but we want them to be on the builders. I'll clarify.